test/common/common/token_bucket_impl_test.cc

```diff
@@ -110,4 +112,63 @@ TEST_F(TokenBucketImplTest, ConsumeAndNextToken) {
   EXPECT_EQ(time_to_next_token, token_bucket.nextTokenAvailable());
 }
 
+TEST_F(TokenBucketImplTest, OptimizeConsume) {
+  uint64_t max_tokens{1000};
+  double tokens_per_sec(500);
+  auto IsImproved = [](uint64_t optimize_access_nums, uint64_t original_access_nums) {
+    return optimize_access_nums <= original_access_nums;
+  };
+
+  auto TestConsume = [](auto&& token_bucket, uint64_t consume_tokens) -> uint64_t {
+    // 测试时间间隔,默认为秒
+    uint64_t time_span{100000};
+    uint64_t access_nums{0};
+
+    uint64_t elapsed_time{0};
+    while (elapsed_time++ <= time_span) {
+      if (token_bucket.consume(consume_tokens, false)) {
```

Comment:
Some things I noticed:
- No calls to `time_system_.advanceTimeWait`.  I think that the time is frozen in these tests.
- It seems that the tests that call IsImproved have optimize_access_nums == original_access_nums, so both the new and old algorithm seem to produce exactly the same results. This seems surprising since we expect the new and old algorithm to behave differently.

It would be helpful to be more specific about the exact results returned by the two algorithms.  IsImproved should use < instead of <= to prove that the cases where an improvement is expected did get an improvement.  Also, is a lower access num with the new algorithm an improvement?  I got the impression earlier that the reason for this change was that there were some inaccuracies in the rate limiter calculation that resulted in the rate limiter providing a rate that is lower than the desired rate but the test is verifying the opposite.

I found that adding `time_system_.advanceTimeWait(std::chrono::milliseconds(1));` before this line results in some differences in the number of tokens that are consumed in the cases where consume_tokens is 400 or 600.  I didn't do the analysis to determine if the new or old behavior is more correct.

Reply:
Thanks！

Yes，IsImproved should use > instead of <= to prove what we talked.

I am not clear about the function of advanceTimeWait，so I don't use that。I now go to research advanceTimeWait and improve the test。

Reply:
This test uses SimulatedTimeSystem which is a test clock and time source which you can advance by hand.  Calls to time_source_.monotonicTime() between calls to advanceTimeWait will return the same value since time doesn't move forward unless you call advanceTimeWait.

The arguments you pass to advanceTimeWait could allow you to move time forward in specific sequences which can be used to demonstrate how the old and new algorithms behave when the time interval between calls to TokenBucketImpl::consume are small.  I think those are the cases where the rounding and accumulation behavior of the two algorithms differ.

Reply:
thanks, I get it! I will update the test!

Reply:
I have get results from the test!

In signle thread case,
Expected equality of these values:
  access_nums
    Which is: 752
  original_access_nums
    Which is: 753
But I have found it In line with expectations。Because，when the current limit exceeds the threshold, the passed qps, that is, access_num is as small as possible, which will be more accurate. But the test shows that the difference is only 1.

I am not sure if this single test needs to be revised！

Reply:
Very sad, the results of my test, the difference between the two is not big, I think this pr may be redundant.

Reply:
Does it make sense to continue with this PR then? It just got assigned to me and I'm not sure of the full context.

Reply:
Hi，thanks CR！

The context is: I want to optimize the implementation of the token bucket. In my use, the native implementation will have some cumulative errors, resulting in inaccurate current limit. Therefore, I have made certain optimizations and found that it is better than the native in use, but uint test shows that the two are basically the same. Now I don’t know if I need to continue.

Reply:
@jmarantz antoniovicente